### PR TITLE
osd/PeeringState: Add logs around can_serve_replica_read() / last_complete_ondisk()

### PR DESCRIPTION
**Pull request by Matan-B**
Few commits which were moved out of #49116 into a separate PR. (For easier QA/CR)

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/osd/PeeringState.cc [hunk 1]
```diff
@@ -3083,6 +3103,34 @@ void PeeringState::proc_replica_log(
   peer_missing[from].claim(std::move(omissing));
 }
 
+void PeeringState::calc_min_last_complete_ondisk() {
+  eversion_t min = last_complete_ondisk;
+  ceph_assert(!acting_recovery_backfill.empty());
```
Comment (ronen-fr) [inline src/osd/PeeringState.cc:3108]: can probably be moved above the previous line (i.e. placed as the 1st line of the function)

File: src/osd/PeeringState.cc [hunk 2]
```diff
@@ -3083,6 +3103,34 @@ void PeeringState::proc_replica_log(
   peer_missing[from].claim(std::move(omissing));
 }
 
+void PeeringState::calc_min_last_complete_ondisk() {
```
Comment (ronen-fr) [inline src/osd/PeeringState.cc:3106]: Can you add a comment describing the idea behind this?

File: src/osd/PeeringState.cc [hunk 3]
```diff
@@ -3083,6 +3103,34 @@ void PeeringState::proc_replica_log(
...
+      continue;
+    }
+    if (peer_last_complete_ondisk.count(pg_shard) == 0) {
+      psdout(20) << " no complete info on: "
+                 << pg_shard << dendl;
+      return;
+    }
+    eversion_t a = peer_last_complete_ondisk[pg_shard];
+    if (a < min) {
+      psdout(20) << " found older version: "
```
Comment (ronen-fr) [inline src/osd/PeeringState.cc:3120]: not much data here apart from 'a' (maybe, at least; add the 'shard'?)

Comment (ronen-fr) [reply src/osd/PeeringState.cc:3120]: (I know it's mostly a moved function. Still...)

**Review by ronen-fr — COMMENTED** [review]
LGTM

**Review by ronen-fr — APPROVED** [review]

File: src/osd/PeeringState.cc [hunk 4]
```diff
@@ -1402,6 +1402,26 @@ bool PeeringState::needs_backfill() const
...
 
+/**
+* Returns whether a particular object can be safely read on this replica
+*/
+bool PeeringState::can_serve_replica_read(const hobject_t &hoid)
+{
+  ceph_assert(!is_primary());
+  eversion_t min_last_complete_ondisk = get_min_last_complete_ondisk();
+  psdout(10) << __func__ << " oid: " << hoid
+             << "min last complete ondisk: " << min_last_complete_ondisk
```
Comment (athanatos) [inline src/osd/PeeringState.cc:1413]: Isn't min_last_complete_ondisk already in the dout prefix as mlcod=?

File: src/osd/PeeringState.cc [hunk 5]
```diff
@@ -3083,6 +3103,35 @@ void PeeringState::proc_replica_log(
...
+*/
+void PeeringState::calc_min_last_complete_ondisk() {
+  ceph_assert(!acting_recovery_backfill.empty());
+  eversion_t min = last_complete_ondisk;
+  for (const auto& pg_shard : acting_recovery_backfill) {
+    if (pg_shard == get_primary()) {
+      continue;
+    }
+    if (peer_last_complete_ondisk.count(pg_shard) == 0) {
+      psdout(20) << " no complete info on: "
```
Comment (athanatos) [inline src/osd/PeeringState.cc:3118]: All of these log lines should have the PeeringState::<method_name> at the start of the log line.

File: src/osd/PeeringState.cc (hunk 4 above)
Comment (Matan-B) [reply src/osd/PeeringState.cc:1413]: Yes, removed.

File: src/osd/PeeringState.cc (hunk 5 above)
Comment (Matan-B) [reply src/osd/PeeringState.cc:3118]: Added "PeeringState::" to dout_prefix

File: src/osd/PeeringState.cc [hunk 6]
```diff
@@ -98,7 +98,8 @@ void PGPool::update(OSDMapRef map)
 
 /*-------------Peering State Helpers----------------*/
 #undef dout_prefix
-#define dout_prefix (dpp->gen_prefix(*_dout))
+#define dout_prefix (dpp->gen_prefix(*_dout)) \
+        << "PeeringState::"
```
Comment (athanatos) [inline src/osd/PeeringState.cc:102]: Hmm, good idea.  How about "PeeringState::" <<  __func__  and remove the __func__ from elsewhere in this file?

Comment (athanatos) [reply src/osd/PeeringState.cc:102]: Note, that won't behave correctly for any uses withing lambdas, so do check for that.

Comment (Matan-B) [reply src/osd/PeeringState.cc]: Updated.

File: src/osd/PeeringState.cc [hunk 7]
```diff
@@ -4556,6 +4556,24 @@ void PeeringState::set_revert_with_targets(
   }
 }
 
+void PeeringState::update_peer_last_complete_ondisk(
+  pg_shard_t fromosd,
+  eversion_t lcod) {
+  psdout(10) << __func__
```
Comment (athanatos) [inline src/osd/PeeringState.cc:4563]: 20

File: src/osd/PeeringState.cc [hunk 8]
```diff
@@ -4556,6 +4556,24 @@ void PeeringState::set_revert_with_targets(
...
+             << " updating peer_last_complete_ondisk"
+             << " of osd: "<< fromosd << " to: "
+             << lcod << dendl;
+  peer_last_complete_ondisk[fromosd] = lcod;
+}
+
+void PeeringState::update_last_complete_ondisk(
+  eversion_t lcod) {
+  psdout(10) << __func__
```
Comment (athanatos) [inline src/osd/PeeringState.cc:4572]: 20

File: src/osd/PeeringState.cc [hunk 9]
```diff
@@ -1402,6 +1402,25 @@ bool PeeringState::needs_backfill() const
...
+/**
+* Returns whether a particular object can be safely read on this replica
+*/
+bool PeeringState::can_serve_replica_read(const hobject_t &hoid)
+{
+  ceph_assert(!is_primary());
+  eversion_t min_last_complete_ondisk = get_min_last_complete_ondisk();
+  if (!pg_log.get_log().has_write_since(
+      hoid, min_last_complete_ondisk)) {
+    psdout(10) << __func__
```
Comment (athanatos) [inline src/osd/PeeringState.cc:1415]: 20

File: src/osd/PeeringState.cc [hunk 10]
```diff
@@ -1402,6 +1402,25 @@ bool PeeringState::needs_backfill() const
...
+               << " can be safely read on this replica" << dendl;
+    return true;
+  } else {
+    psdout(10) << __func__
```
Comment (athanatos) [inline src/osd/PeeringState.cc:1419]: 20

File: src/osd/PeeringState.cc [hunk 11]
```diff
@@ -3102,6 +3102,35 @@ void PeeringState::proc_replica_log(
...
+      psdout(20) << __func__ <<  " no complete info on: "
+                 << pg_shard << dendl;
+      return;
+    }
+    if (peer_last_complete_ondisk[pg_shard] < min) {
+      min = peer_last_complete_ondisk[pg_shard];
+    }
+  }
+  if (min != min_last_complete_ondisk) {
+    psdout(10) << __func__ << " last_complete_ondisk is "
```
Comment (athanatos) [inline src/osd/PeeringState.cc]: 20

**Review by athanatos — CHANGES_REQUESTED** [review]
LGTM other than switching the logging from 10->20.  Those lines are printed at a rate proportional to incoming IO, so they need to be at 20.  As a suggestion for the future, your commits combine moving a method between two files, updating the loop structure, and adding logging.  Separating those three things into distinct commits makes it much simpler to review since I wouldn't have to go line-by-line to identify your changes.

**Review by athanatos — APPROVED** [review]
Needs a rados suite test run.

**Comment by ljflores** [comment]
@Matan-B I see two repeated failures from the testing batch that I believe are related to this PR. Can you take a look?

https://pulpito.ceph.com/yuriw-2023-02-07_16:11:40-rados-wip-yuri10-testing-2023-02-06-0846-distro-default-smithi/7154278/
```
2023-02-10T11:48:28.167 INFO:tasks.workunit.client.0.smithi072.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  grep 'pg[[]1.0.*backfilling.*update_calc_stats degraded ' td/osd-backfill-stats/osd.1.log
2023-02-10T11:48:28.167 INFO:tasks.workunit.client.0.smithi072.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  grep -F ' [1,2]/[1] '
2023-02-10T11:48:28.168 INFO:tasks.workunit.client.0.smithi072.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  grep -v est
2023-02-10T11:48:28.168 INFO:tasks.workunit.client.0.smithi072.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  head -1
2023-02-10T11:48:28.168 INFO:tasks.workunit.client.0.smithi072.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  sed 's/.* \([0-9]*\)$/\1/'
2023-02-10T11:48:28.175 INFO:tasks.workunit.client.0.smithi072.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:99: check:  FIRST=
2023-02-10T11:48:28.176 INFO:tasks.workunit.client.0.smithi072.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:100: check:  '[' -n '' ']'
2023-02-10T11:48:28.176 INFO:tasks.workunit.client.0.smithi072.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:100: check:  return 1
2023-02-10T11:48:28.176 INFO:tasks.workunit.client.0.smithi072.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd-backfill/osd-backfill-stats.sh:316: TEST_backfill_down_out:  return 1
```

https://pulpito.ceph.com/yuriw-2023-02-07_16:11:40-rados-wip-yuri10-testing-2023-02-06-0846-distro-default-smithi/7154355/
```
2023-02-10T11:17:54.535 INFO:tasks.workunit.client.0.smithi052.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  grep 'pg[[]1.0.*recovering.*update_calc_stats degraded ' td/osd-recovery-stats/osd.1.log
2023-02-10T11:17:54.535 INFO:tasks.workunit.client.0.smithi052.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  grep -F ' [1,3,2]/[1,NONE,2]p'
2023-02-10T11:17:54.535 INFO:tasks.workunit.client.0.smithi052.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  grep -v est
2023-02-10T11:17:54.535 INFO:tasks.workunit.client.0.smithi052.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  head -1
2023-02-10T11:17:54.535 INFO:tasks.workunit.client.0.smithi052.stderr:///home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  sed 's/.* \([0-9]*\)$/\1/'
2023-02-10T11:17:54.554 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:89: check:  FIRST=
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:90: check:  below_margin 200
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:46: below_margin:  local -i check=200
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:47: below_margin:  shift
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:48: below_margin:  local -i target=
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:50: below_margin:  return 1
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:90: check:  return 1
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:162: do_recovery_out1:  return 1
2023-02-10T11:17:54.555 INFO:tasks.workunit.client.0.smithi052.stderr:/home/ubuntu/cephtest/clone.client.0/qa/standalone/osd/osd-recovery-stats.sh:177: TEST_recovery_erasure_out1:  return 1
```

There are two different workunits that failed here, but they failed in a very similar way. It's not super obvious how it might be caused by this PR, as I see you are pretty much only adding extra logging. But both workunits seem to use `update_calc_stats`, which I see has been modified in this PR.

**Comment by ljflores** [comment]
I have asked @yuriw to remove the PR from the batch and rerun those two tests to verify. If they turn out unrelated, I'll let you know.

This is the Trello card BTW: https://trello.com/c/FzMz7O3S/1683-wip-yuri10-testing-2023-02-06-0846-old-wip-yuri10-testing-2023-02-06-0809

**Comment by Matan-B** [comment]
Thank you Laura!
Looks like `osd-recovery-stats.sh`/`osd-backfill-stats.sh` should be adjusted as well since specfic log lines (which were changed here) are being grepped. I'll address it.
```
FIND_UPACT='grep "pg[[]${PG}.*recovering.*update_calc_stats " $log | tail -1 | sed "s/.*[)] \([[][^ p]*\).*$/\1/"'
FIND_FIRST='grep "pg[[]${PG}.*recovering.*update_calc_stats $which " $log | grep -F " ${UPACT}${addp}" | grep -v est | head -1 | sed "s/.* \([0-9]*\)$/\1/"'
FIND_LAST='grep "pg[[]${PG}.*recovering.*update_calc_stats $which " $log | tail -1 | sed "s/.* \([0-9]*\)$/\1/"'
```

**Comment by Matan-B** [comment]
Fix attempt: 72227ccff44b520b804f528e9d76e21a0d39ab27
Re-testing failing jobs:
https://pulpito.ceph.com/matan-2023-02-19_08:52:59-rados:standalone-wip-matanb-lrod-logs-testing-distro-default-smithi/

**Comment by ljflores** [comment]
@Matan-B message me in gchat if you want it back in one of Yuri's batches

**Comment by Matan-B** [comment]
https://pulpito.ceph.com/matan-2023-03-07_15:15:56-rados:standalone-wip-matanb-lrod-logs-v2-distro-default-smithi/

**Comment by Matan-B** [comment]
jenkins test api

**Comment by ljflores** [comment]
https://pulpito.ceph.com/?branch=wip-yuri4-testing-2023-03-15-1418

Failures, unrelated:
    1. https://tracker.ceph.com/issues/58946
    2. https://tracker.ceph.com/issues/56393
    3. https://tracker.ceph.com/issues/59123
    4. https://tracker.ceph.com/issues/58585
    5. https://tracker.ceph.com/issues/58560
    6. https://tracker.ceph.com/issues/59127

Details:
    1. cephadm: KeyError: 'osdspec_affinity' - Ceph - Orchestrator
    2. thrash-erasure-code-big: failed to complete snap trimming before timeout - Ceph - RADOS
    3. Timeout opening channel - Infrastructure
    4. rook: failed to pull kubelet image - Ceph - Orchestrator
    5. test_envlibrados_for_rocksdb.sh failed to subscribe to repo - Infrastructure
    6. Job that normally complete much sooner last almost 12 hours - Infrastructure

**Pull request merged**